Services/UI/classes/class.ilUIFilterService.php

```diff
@@ -0,0 +1,216 @@
+			$this->request->getAction($base_action, self::CMD_COLLAPSE),
+			$this->request->getAction($base_action, self::CMD_APPLY),
+			$this->request->getAction($base_action, self::CMD_RESET),
+			$inputs_with_session_data,
+			$is_input_initially_rendered_with_session,
+			$is_activated,
+			$is_expanded);
+
+		// handle apply command
+		$filter = $this->handleApply($filter_id, $filter);
+
+		return $filter;
+
+	}
+
+	/**
+	 * Get data
+	 *
+	 * @param \ILIAS\UI\Component\Input\Container\Filter\Standard $filter
+	 * @return array|null
+	 */
+	public function getData(\ILIAS\UI\Component\Input\Container\Filter\Standard $filter)
+	{
+		$result = null;
+		if (in_array($this->request->getFilterCmd(),
+				[self::CMD_APPLY, self::CMD_TOGGLE_ON, self::CMD_EXPAND, self::CMD_COLLAPSE]) && $this->request->isPost()) {
+			$filter = $this->request->getFilterWithRequest($filter);
+			$result = $filter->getData();
+		}
+		return $result;
+	}
+
+	/**
+	 * Write filter status to session (filter activated/expanded, inputs being rendered or not)
+	 * @param string $filter_id
+	 * @param array $inputs
+	 */
+	protected function writeFilterStatusToSession($filter_id, $inputs)
+	{
+		if ($this->request->getFilterCmd() == self::CMD_TOGGLE_ON) {
+			$this->session->writeActivated($filter_id, true);
+		}
+
+		if ($this->request->getFilterCmd() == self::CMD_TOGGLE_OFF) {
+			$this->session->writeActivated($filter_id, false);
+		}
+
+		if ($this->request->getFilterCmd() == self::CMD_EXPAND) {
+			$this->session->writeExpanded($filter_id, true);
+		}
+
+		if ($this->request->getFilterCmd() == self::CMD_COLLAPSE) {
+			$this->session->writeExpanded($filter_id, false);
+		}
+
+		if ($this->request->getFilterCmd() == self::CMD_APPLY)
+		{
+			foreach ($inputs as $input_id => $i)
+			{
+				if ($this->request->isInputRendered($input_id))
+				{
+					$this->session->writeRendered($filter_id, $input_id, true);
+				}
+				else
+				{
+					$this->session->writeRendered($filter_id, $input_id, false);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Handle reset command
+	 *
+	 * @param string $filter_id
+	 */
+	protected function handleReset(string $filter_id)
+	{
+		// clear session, if reset is pressed
+		if ($this->request->getFilterCmd() == self::CMD_RESET)
+		{
+			$this->session->reset($filter_id);
+		}
+	}
+
+
+	/**
+	 * Handle apply command
+	 *
+	 * @param string $filter_id
+	 * @param \ILIAS\UI\Component\Input\Container\Filter\Standard $filter
+	 * @return \ILIAS\UI\Component\Input\Container\Filter\Standard
+	 */
+	protected function handleApply(string $filter_id, \ILIAS\UI\Component\Input\Container\Filter\Standard $filter): \ILIAS\UI\Component\Input\Container\Filter\Standard
+	{
+		if ($this->request->getFilterCmd() == self::CMD_APPLY &&
+			$this->request->isPost())
+		{
+			$filter = $this->request->getFilterWithRequest($filter);
+			foreach ($filter->getInputs() as $input_id => $i)
+			{
+				$this->session->writeValue($filter_id, $input_id, $i->getValue());
+			}
```

Comment:
Yes, I see. Would it be ok, to make the non numeric keys in the submitted array mandatory? Like in the example?
```
[
		"title" => $title_input,
		"select" => $select,
		"with_def" => $with_def,
		"init_hide" => $init_hide,
	]
```
We could test for these keys. This would fix this issue for now.

Reply:
No, there just are no guarantees here. This will work for now, but since the inputs are not meant to be read in this way this could well change. This is what I meant with "against the grain" in my other comment.